STM32L4: Add support of the new USB Device API

bcostm opened the pull request:
### Description

Enable new USB Device API on the following STM32L4 platforms:
- NUCLEO_L496ZG
- NUCLEO_L496ZG_P
- NUCLEO_L4R5ZI
- DISCO_L496AG
- DISCO_L475VG_IOT01A
- DISCO_L476VG

The same USB HAL patch as done for STM32F4 and STM32F2 has been applied (see PRs #7322 and #8583)

### Tests
- [x] tests-usb_device-basic tests are OK on ARM, IAR and GCC_ARM
- [ ] tests-usb_device-serial tests are FAIL

### Pull request type

    [ ] Fix
    [ ] Refactor
    [X] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

bcostm committed:
STM32L4 USB: add patch in CubeL4 hal driver

bcostm committed:
STM32L4 USB: add pins definition for STM32L496/L4R5 MCUs

adbridge commented:
@c1728p9 This review has been waiting for 7 days could you please take a look asap

c1728p9 commented on targets/TARGET_STM/TARGET_STM32L4/device/stm32l4xx_hal_pcd.c:
```diff
@@ -311,10 +310,10 @@ __weak void HAL_PCD_MspDeInit(PCD_HandleTypeDef *hpcd)
 HAL_StatusTypeDef HAL_PCD_Start(PCD_HandleTypeDef *hpcd)
 { 
-  __HAL_LOCK(hpcd); 
+  // MBED PATCH __HAL_LOCK(hpcd); 
```
Why was the lock removed?

bcostm replied:
It has been removed in the F4 file so I just did the same...

c1728p9 commented on targets/TARGET_STM/TARGET_STM32L4/device/stm32l4xx_hal_pcd.c:
```diff
@@ -162,11 +163,9 @@ HAL_StatusTypeDef HAL_PCD_Init(PCD_HandleTypeDef *hpcd)
-    hpcd->EPLock[index].Lock = HAL_UNLOCKED;
-
+    for (index = 0; index < hpcd->Init.dev_endpoints ; index++) { // MBED PATCH
+      hpcd->EPLock[index].Lock = HAL_UNLOCKED;
```
Is this just cosmetic?

bcostm replied:
Yes

0xc0170 commented:
Note: we are currently having high load on the CI. It might take some time. 

@c1728p9 this is not scheduled for 5.11 , is it?

0xc0170 commented:
Info: This PR has been re-bundled into a new rollup PR (#8768).

No further work is needed here, as once that PR is merged, this PR will also be closed and marked as merged.
If any more commits are made in this PR, this PR will remain open and have to go through CI on its own.

bcostm commented:
OK. I will not add more commits here.

I will send a new PR later to align the PCD functions (Init, Start, Stop) in the F2, F4, F7 and L4 families.

cmonr commented:
@bcostm It looks like both PRs in the last rollup PR had issues. Please take a look at the L4-related build failures: https://github.com/ARMmbed/mbed-os/pull/8768#issuecomment-439357325

bcostm commented:
Rebased. I don't understand the build errors with the rollup PR for the L4 boards related to USB ? Because these boards don't have the USBDEVICE macro in targets.json. How is it possible ?

cmonr commented:
@bcostm It's possible it was an interpretation error on my part.

Restarting CI.

screamerbg commented:
@0xc0170 @cmonr What's the issue with this PR and CI? We're very close to feature freeze for Mbed OS 5.11 and we need to get this in (assuming it's CI issues)

bcostm commented:
I have reproduced the build error and found the root cause. I'll send a new commit to fix it.

bcostm committed:
STM32L4 USB: move HAL_PCD_EP_Abort function

This function is for USB_OTG_FS devices only. Move it in the correct place (in "#ifdef USB_OTG_FS" area).

0xc0170 commented:
> @0xc0170 @cmonr What's the issue with this PR and CI? We're very close to feature freeze for Mbed OS 5.11 and we need to get this in (assuming it's CI issues)

This answers my question I asked a week ago above - this is targeting 5.11. 

There were build problems and CI issues, we migrated to the new CI so improve. We will start CI soon (rollup PR and kvstore need to pass first , they should finish within few hours).

0xc0170 commented:
CI started

0xc0170 commented:
This is the build failure:

```
21:11:16 [MTS_DRAGONFLY_F411RE:ARM] # mbed-os-example-cellular MTS_DRAGONFLY_F411RE ARM

21:11:16 [MTS_DRAGONFLY_F411RE:ARM] UndefinedParameter: Attempt to override undefined parameter 'nsapi.default-cellular-plmn' in 'application[*]'
```

I do not see any change related to this. @jarvte CAn you review the failure? I looked at the cellular example, there were recent commits (related to this failure) ? Did anything break?

jarvte commented:
> This is the build failure:
> 
> ```
> 21:11:16 [MTS_DRAGONFLY_F411RE:ARM] # mbed-os-example-cellular MTS_DRAGONFLY_F411RE ARM
> 
> 21:11:16 [MTS_DRAGONFLY_F411RE:ARM] UndefinedParameter: Attempt to override undefined parameter 'nsapi.default-cellular-plmn' in 'application[*]'
> ```
> I do not see any change related to this. @jarvte CAn you review the failure? I looked at the cellular example, there were recent commits (related to this failure) ? Did anything break?

Can you try again? We changed cellular example master to use mbed-os master on friday. 
Mbed master has 'nsapi.default-cellular-plmn'. Our ci works fine and it uses master. This PR is not a reason for failure.

cmonr commented:
> Our ci works fine and it uses master. This PR is not a reason for failure.

It's possible that this will need a rebase then if the feature branch does not have `nsapi.default-cellular-plmn`.

bcostm commented:
Can someone rebase the feature-hal-spec-usb-device branch ? I'll rebase my branch afterwards.

cmonr commented:
Rebased. All of the conflicts occurred with `targets/targets.json`.
A new macro was added (`USB_STM_HAL`), in master, but not sure if it should be in this branch. Feel free to correct that if needed with this PR.

cmonr commented:
@bcostm When you get a chance, the base branch has been rebased.

bcostm commented:
Rebased.
tests-usb_device-basic tests are OK
tests-usb_device-serial tests are FAIL --> I need ARM USB experts to help solving this issue at least on one ST board. I'll recheck all other boards afterwards. Thanks

cmonr commented:
@c1728p9 Mind helping out?

screamerbg commented:
@c1728p9 Could you help?

cmonr commented:
@screamerbg I suspect he's getting tied up with release tasks, which should be wrapping up within a day or two.

c1728p9 commented:
I rebased the USB feature branch so it is up to date. @bcostm you'll need to rebase this and then testing can be restarted.

bcostm commented:
Thanks @c1728p9 
I have rebased my branch but Serial tests are still fail (tested on DISCO_L475VG_IOT and DISCO_L496AG platforms). Basic USB tests are OK.

c1728p9 commented:
@bcostm can you post your test results for Serial? Are you seeing the same type Serial test failures on the `NUCLEO_F207ZG` as you are on the other boards? Also, what OS are you testing on?

bcostm commented:
Yes I see the same failure on all boards. The test `CDC USB reconnect ` is FAIL and other tests are skipped. I use Windows7.

bcostm commented:
Can you confirm if the Serial tests are OK on your side using Windows ?

0xc0170 commented:
CI started in the meantime

NirSonnenschein commented:
Hi @bcostm , 
There seem to be some build issues when running the CI: 
15:03:02 [DISCO_F746NG:ARM] [Error] USBPhy_STM32.cpp@518,0:  #20: identifier "HAL_PCD_EP_Abort" is undefined
15:03:02 [DISCO_F746NG:ARM] Failed to build library

please take a look.

bcostm commented:
There was a problem in targets.json file. The USBDEVICE was activated on the DISCO_F746NG and also on the MTS_DRAGONFLY_L471QG platforms (?). Should be ok now.

cmonr commented:
_sigh_

And now the feature branch itself needs a rebase because of Travis CI fixes...

@bcostm Will start in a bit once the feature branch is sorted out. The TLDR is that the feature branch needs this PR (https://github.com/ARMmbed/mbed-os/pull/9289) and a fix for an astyle simplification which I'm resolving now (ref: https://github.com/ARMmbed/mbed-os/pull/9287#issuecomment-452724407).

cmonr commented:
Note: https://github.com/ARMmbed/mbed-os/pull/7976#issuecomment-453383516

0xc0170 commented:
@c1728p9 Can you help rebasing the branch please?

bcostm commented:
Rebased

0xc0170 commented:
I restarted travis failure, and will schedule CI as soon as CI is back from updates

0xc0170 commented:
CI restarted

0xc0170 commented:
`[DEBUG] Output: "./usb/device/targets/TARGET_STM/USBPhy_STM32.cpp", line 518: Error:  #20: identifier "HAL_PCD_EP_Abort" is undefined` found in the build logs, please review

bcostm commented:
I don't understand what happens here ??? It is the same error as 10 days ago which I already resolved. Now it comes back again after a Rebase ? @jeromecoutant can you please have a look ?

jeromecoutant commented on targets/targets.json:
```diff
@@ -3777,6 +3777,7 @@
             "TRNG",
             "FLASH",
             "QSPI",
+            "USBDEVICE",
```
```suggestion
```

bcostm committed:
STM32L4 USB: enable USBDEVICE

jeromecoutant commented:
CI restart :-)

0xc0170 commented:
CI restarted

0xc0170 commented:
Test restarted (CI agent failures)
